### PR TITLE
Applying from branch Ingress

### PR DESCRIPTION
**Pull request by tanyaveksler**

### CODE REVIEW

File: tests/calico_testcases/example_policies/expected_output/testcase15_with_ingress_connectivity_map.txt [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+src_ns: [kube-system] src_pods: [*] dst_ns: [kube-system] dst_pods: [calico-node] conn: TCP {'dst_ports': '210', 'paths': '/abc(/(([\\--9A-Z_abce-z]|d([\\--9A-Z_a-df-z]|e([\\--9A-Z_a-eg-z]|f[\\-.0-9A-Z_a-z])))[\\--9A-Z_a-z]*|d(e(f/)?)?))?', 'hosts': 'first.bar.com'}
+src_ns: [kube-system] src_pods: [*] dst_ns: [kube-system] dst_pods: [file-plugin-7bfb8b69bf] conn: TCP {'dst_ports': '80', 'paths': '/abc/def(/[\\--9A-Z_a-z]+)?', 'hosts': 'first.bar.com'}
+src_ns: [kube-system] src_pods: [*] dst_ns: [kube-system] dst_pods: [kube-dns-amd64-d66bf76db] conn: TCP {'dst_ports': '213', 'paths': '/[\\--9A-Z_a-z]*'}
+src_ns: [kube-system] src_pods: [*] dst_ns: [kube-system] dst_pods: [kube-fluentd] conn: TCP {'dst_ports': '80', 'paths': '/xyz/[\\--9A-Z_a-z]+', 'hosts': 'second.bar.com'}
+src_ns: [kube-system] src_pods: [*] dst_ns: [kube-system] dst_pods: [storage-watcher-8494b4b8bb] conn: TCP {'dst_ports': '102', 'paths': '/xyz', 'hosts': 'second.bar.com'}
+src_ns: [kube-system] src_pods: [calico-kube-controllers-7694668c77] dst_ns: [default,kube-system,vendor-system] dst_pods: [*] conn: All connections
```
Comment (adisos) [inline tests/calico_testcases/example_policies/expected_output/testcase15_with_ingress_connectivity_map.txt:12]: I am not sure that I understand this expected output... I'd expect Ingress to affect only on the connectivity from the ingress-controller pod..

**Review by adisos — COMMENTED** [review]
Added some comment

File: network-config-analyzer/NetworkConfigQuery.py [hunk 2]
```diff
@@ -584,6 +622,19 @@ def disjoint_referenced_ip_blocks(self):
...
+        Clone config without ingress policies
+        :param config: the config to clone
+        :return: resulting config without ingress policies
+        """
+        config_without_ingress = config.clone_without_policies(config.name)
+        for policy in config.policies.values():
+            if not isinstance(policy, IngressPolicy):  # ignoring ingress policies
+                config_without_ingress.append_policy_to_config(policy)
+        return config_without_ingress
+
```
Comment (adisos) [inline network-config-analyzer/NetworkConfigQuery.py:637]: Maybe would be more efficient to first check if Ingress policies exist in the config, and if not - just return the input config..

File: network-config-analyzer/CalicoNetworkPolicy.py [hunk 3]
```diff
@@ -75,7 +75,8 @@ def __init__(self, name, namespace):
         self.order = None  # None means infinity here
 
     def __eq__(self, other):
-        return super().__eq__(other) and self.order == other.order
+        return super().__eq__(other) and isinstance(other, CalicoNetworkPolicy) and \
+            self.order == other.order
```
Comment (adisos) [inline network-config-analyzer/CalicoNetworkPolicy.py:79]:
```suggestion
        return isinstance(other, CalicoNetworkPolicy) and super().__eq__(other) and \
            self.order == other.order
```
Can avoid calling  super().__eq__ if the other policy type does not match.

File: network-config-analyzer/NetworkConfigQuery.py (hunk 2 above)
Comment (tanyaveksler) [reply network-config-analyzer/NetworkConfigQuery.py]: Done

File: tests/calico_testcases/example_policies/expected_output/testcase15_with_ingress_connectivity_map.txt (hunk 1 above)
Comment (tanyaveksler) [reply tests/calico_testcases/example_policies/expected_output/testcase15_with_ingress_connectivity_map.txt:12]: I think this is another way to define the limited connectivity from the ingress controller pods: limit connectivity from all pods in kube-system namespace, and add 'All connections' from all but ingress controller pods in this namespace.

Comment (adisos) [reply tests/calico_testcases/example_policies/expected_output/testcase15_with_ingress_connectivity_map.txt:12]: Right, ok

**Review by adisos — COMMENTED** [review]
We still have some runtime gap: on test tests/calico_testcases/example_policies/testcase15-ports/testcase15-with-ingress-scheme.yaml , the expected runtime is 28.96 and the actual runtime was 113.02, from the last push.

File: tests/calico_testcases/example_policies/expected_output/testcase15_with_ingress_connectivity_map.txt [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+src_ns: [kube-system] src_pods: [tiller-deploy-5c45c9966b] dst_ns: [default,kube-system,vendor-system] dst_pods: [*] conn: All connections
+src_ns: [kube-system] src_pods: [vpn-858f6d9777] dst_ns: [default,kube-system,vendor-system] dst_pods: [*] conn: All connections
+
+final fw rules for query: connectivity_map, config: global-simple:
+src_ns: [kube-system] src_pods: [app=keepalived-watcher] dst_ns: [kube-system] dst_pods: [!has(app)] conn: TCP 200-250
+src_ns: [kube-system] src_pods: [app=keepalived-watcher] dst_ns: [vendor-system] dst_pods: [*] conn: TCP 200-250
+
+final fw rules for query: connectivity_map, config: global-simple-with-ingress:
+src_ns: [kube-system] src_pods: [app=keepalived-watcher] dst_ns: [kube-system] dst_pods: [calico-node] conn: TCP {'dst_ports': '210', 'paths': '/abc(/(([\\--9A-Z_abce-z]|d([\\--9A-Z_a-df-z]|e([\\--9A-Z_a-eg-z]|f[\\-.0-9A-Z_a-z])))[\\--9A-Z_a-z]*|d(e(f/)?)?))?', 'hosts': 'first.bar.com'}
+src_ns: [kube-system] src_pods: [app=keepalived-watcher] dst_ns: [kube-system] dst_pods: [kube-dns-amd64-d66bf76db] conn: TCP {'dst_ports': '213', 'paths': '/[\\--9A-Z_a-z]*'}
```
Comment (adisos) [inline tests/calico_testcases/example_policies/expected_output/testcase15_with_ingress_connectivity_map.txt:31]: Why in this part of the expected output it seems that the Ingress policy affects on allowed connections from a pod which is not the ingress controller?

Comment (tanyaveksler) [reply tests/calico_testcases/example_policies/expected_output/testcase15_with_ingress_connectivity_map.txt:31]: In tests/calico_testcases/example_podlist/service_and_ingress_resources/services_list.yaml this is how I defined ingress controller:
`apiVersion: v1
kind: Service
metadata:
  name: ingress-nginx-controller
  namespace: kube-system
spec:
  type: NodePort
  selector:
    **app.kubernetes.io/name: ingress-nginx**
    **app: keepalived-watcher**
  ports:
    - appProtocol: https
      name: https-webhook
      port: 443
      protocol: TCP
      targetPort: webhook`
So the pods with app: keepalived-watcher are actually the ingress controller pods.

**Review by adisos — APPROVED** [review]

**Pull request merged**